sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala

```diff
@@ -454,6 +478,28 @@ case class BroadcastHashJoinExec(
     val (matched, checkCondition, _) = getJoinCondition(ctx, input)
     val numOutput = metricTerm(ctx, "numOutputRows")
 
+    if (isNullAwareAntiJoin) {
+      require(leftKeys.length == 1, "leftKeys length should be 1")
+      require(rightKeys.length == 1, "rightKeys length should be 1")
+      require(joinType == LeftAnti, "joinType must be LeftAnti.")
+      require(buildSide == BuildRight, "buildSide must be BuildRight.")
+      require(SQLConf.get.nullAwareAntiJoinOptimizeEnabled,
+        "nullAwareAntiJoinOptimizeEnabled must be on for null aware anti join optimize.")
+      require(checkCondition == "", "null aware anti join optimize condition should be empty.")
+
+      if (broadcastRelation.value.inputEmpty) {
+        return s"""
+           |// singleColumn NAAJ inputEmpty(true) accept all
+           |$numOutput.add(1);
+           |${consume(ctx, input)}
+         """.stripMargin
+      } else if (broadcastRelation.value.anyNullKeyExists) {
+        return s"""
+           |// singleColumn NAAJ inputEmpty(false) anyNullKeyExists(true) reject all
```

Comment:
So there are only 3 special changes for NAAJ on the codegen path, and in total just 5 lines. If we can further unify the code for `uniqueKeyCodePath` and not unique, there will be only 2 special changes for NAAJ.

The `doExecute` is fine as there is a big `if-else` at the beginning to diverge between NAAJ and normal anti join. I don't think it hurts readability too much.

So I think it's better to reuse BHJ, as a new node has a large maintenance cost. The tradeoff is between the readability of BHJ and maintenance cost of a new node, but as I explained I think the readability of BHJ won't be hurt too much.

@maryannxue @agrawaldevesh @maropu @viirya what do you think?

Reply:
@cloud-fan do you know of any history about why hashedRelation was filter out anyNullKey rows?
I mean in the future, to support multi-column NAAJ, I need those  anyNullKey rows still kept inside hashedRelation, is it ok to do this?

See. code

```

while (input.hasNext) {
      val row = input.next().asInstanceOf[UnsafeRow]
      numFields = row.numFields()
      val key = keyGenerator(row)
      if (!key.anyNull) { // Here
        val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
        val success = loc.append(
          key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
          row.getBaseObject, row.getBaseOffset, row.getSizeInBytes)
        if (!success) {
          binaryMap.free()
          // scalastyle:off throwerror
          throw new SparkOutOfMemoryError("There is not enough memory to build hash map")
          // scalastyle:on throwerror
        }

```

Reply:
If the readability won't be bad, I'm ok to reuse BHJ. I support new node only if reuse of BHJ results in bad code hard to read or maintain.

Reply:
I guessed that, since there was no null aware join in past days, so there are no "NullAware HashedRelation".

Reply:
> @maryannxue @agrawaldevesh @maropu @viirya what do you think?

Thanks for the empirical analysis @cloud-fan ... that always helps to seal an argument :-) 

I do agree that introducing a new node is sometimes more pain than worth, so I would be equally be okay with just modifying BHJ. I think they are both fine choices and I would leave it to @leanken to make a call. The ideal scenario would be to have minimal code diff in BHJ and introducing a new node is only a fallback if the diff cannot be reduced to our satisfaction.

Either way, we ought to decide one way or the other soon since the PR is becoming slightly confusing with both approaches co-existing :-)

Are there any performance benchmarks that we should be re-running for BHJ to ensure no regression ?

Reply:
done using BHJ as default and only implementation